clippy_lints/src/missing_inline.rs

```diff
@@ -0,0 +1,171 @@
+//   except according to those terms.
+//
+
+use rustc::hir;
+use rustc::lint::*;
+use syntax::ast;
+use syntax::codemap::Span;
+
+/// **What it does:** it lints if an exported function, method, trait method with default impl,
+/// or trait method impl is not `#[inline]`.
+///
+/// **Why is this bad?** In general, it is not. Functions can be inlined across
+/// crates when that's profitable as long as any form of LTO is used. When LTO is disabled,
+/// functions that are not `#[inline]` cannot be inlined across crates. Certain types of crates
+/// might intend for most of the methods in their public API to be able to be inlined across
+/// crates even when LTO is disabled. For these types of crates, enabling this lint might make sense.
+/// It allows the crate to require all exported methods to be `#[inline]` by default, and then opt
+/// out for specific methods where this might not make sense.
+///
+/// **Known problems:** None.
+declare_clippy_lint! {
+    pub MISSING_INLINE_IN_PUBLIC_ITEMS,
+    restriction,
+    "detects missing #[inline] attribute for public callables (functions, trait methods, methods...)"
+}
+
+pub struct MissingInline {}
+
+impl ::std::default::Default for MissingInline {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MissingInline {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn check_missing_inline_attrs(&self, cx: &LateContext,
+                                  attrs: &[ast::Attribute], sp: Span, desc: &'static str) {
+        // If we're building a test harness, FIXME: is this relevant?
+        // if cx.sess().opts.test {
+        //    return;
+        // }
+
+        let has_inline = attrs
+            .iter()
+            .any(|a| a.name() == "inline" );
+        if !has_inline {
+            cx.span_lint(
+                MISSING_INLINE_IN_PUBLIC_ITEMS,
+                sp,
+                &format!("missing `#[inline]` for {}", desc),
+            );
+        }
+    }
+}
+
+impl LintPass for MissingInline {
+    fn get_lints(&self) -> LintArray {
+        lint_array![MISSING_INLINE_IN_PUBLIC_ITEMS]
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for MissingInline {
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, it: &'tcx hir::Item) {
+        if !cx.access_levels.is_exported(it.id) {
+            return;
+        }
+        match it.node {
+            hir::ItemFn(..) => {
+                // ignore main()
+                if it.name == "main" {
+                    let def_id = cx.tcx.hir.local_def_id(it.id);
+                    let def_key = cx.tcx.hir.def_key(def_id);
+                    if def_key.parent == Some(hir::def_id::CRATE_DEF_INDEX) {
+                        return;
+                    }
+                }
+                let desc = "a function";
+                self.check_missing_inline_attrs(cx, &it.attrs, it.span, desc);
+            },
+            hir::ItemTrait(ref _is_auto, ref _unsafe, ref _generics,
```

Comment:
Add a comment here that we're not using https://doc.rust-lang.org/nightly/nightly-rustc/rustc/lint/trait.LateLintPass.html#method.check_trait_item because we need to check if the trait is exported